db/range_tombstone_fragmenter.cc

```diff
@@ -195,107 +198,143 @@ void FragmentedRangeTombstoneList::FragmentTombstones(
 }
 
 FragmentedRangeTombstoneIterator::FragmentedRangeTombstoneIterator(
-    const FragmentedRangeTombstoneList* tombstones,
+    const FragmentedRangeTombstoneList* tombstones, SequenceNumber snapshot,
     const InternalKeyComparator& icmp)
-    : tombstone_cmp_(icmp.user_comparator()),
-      icmp_(&icmp),
+    : tombstone_start_cmp_(icmp.user_comparator()),
+      tombstone_end_cmp_(icmp.user_comparator()),
       ucmp_(icmp.user_comparator()),
-      tombstones_(tombstones) {
+      tombstones_(tombstones),
+      snapshot_(snapshot) {
   assert(tombstones_ != nullptr);
   pos_ = tombstones_->end();
   pinned_pos_ = tombstones_->end();
 }
 
 FragmentedRangeTombstoneIterator::FragmentedRangeTombstoneIterator(
     const std::shared_ptr<const FragmentedRangeTombstoneList>& tombstones,
-    const InternalKeyComparator& icmp)
-    : tombstone_cmp_(icmp.user_comparator()),
-      icmp_(&icmp),
+    SequenceNumber snapshot, const InternalKeyComparator& icmp)
+    : tombstone_start_cmp_(icmp.user_comparator()),
+      tombstone_end_cmp_(icmp.user_comparator()),
       ucmp_(icmp.user_comparator()),
       tombstones_ref_(tombstones),
-      tombstones_(tombstones_ref_.get()) {
+      tombstones_(tombstones_ref_.get()),
+      snapshot_(snapshot) {
   assert(tombstones_ != nullptr);
   pos_ = tombstones_->end();
+  seq_pos_ = tombstones_->seq_end();
   pinned_pos_ = tombstones_->end();
+  pinned_seq_pos_ = tombstones_->seq_end();
 }
 
 void FragmentedRangeTombstoneIterator::SeekToFirst() {
   pos_ = tombstones_->begin();
+  seq_pos_ = tombstones_->seq_begin();
```

Comment:
`SeekToFirst`, `SeekToLast`, `Next`, and `Prev` look like they're only intended for flush/compaction. should we assert non-one-time-use and snapshot is kMaxSeqnum in such places?

Reply:
That's a good idea; I'll do that.

Reply:
As discussed offline, these functions are used during tests to make it easier for the reader to identify all the tombstone fragments in a list. Since those functions are also used in tests for snapshot and one-time-use, the assertions would have to be written to not fire in those cases. I'll revisit this comment in the cleanup alluded to [here](https://github.com/facebook/rocksdb/pull/4632#discussion_r233657664).